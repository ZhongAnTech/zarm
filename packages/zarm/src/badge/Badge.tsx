import * as React from 'react';
import classnames from 'classnames';
import { ConfigContext } from '../n-config-provider';
import type { BaseBadgeProps } from './interface';

export type BadgeProps = BaseBadgeProps & React.HTMLAttributes<HTMLSpanElement>;

const Badge = React.forwardRef<HTMLSpanElement, BadgeProps>((props, ref) => {
  const { className, theme, shape, text, children, ...restProps } = props;
  const badgeRef = ref || React.createRef<HTMLSpanElement>();

  const { prefixCls: globalPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-badge`;

  const cls = classnames(prefixCls, className, {
    [`${prefixCls}--${theme}`]: !!theme,
    [`${prefixCls}--${shape}`]: shape,
    [`${prefixCls}--sup`]: !!children,
  });

  return (
    <span ref={badgeRef} className={cls} {...restProps}>
      {children}
      <sup className={`${prefixCls}__content`}>{shape !== 'dot' && text}</sup>
    </span>
  );
});

Badge.displayName = 'Badge';

Badge.defaultProps = {
  shape: 'dot',
  theme: 'danger',
};

export default Badge;
