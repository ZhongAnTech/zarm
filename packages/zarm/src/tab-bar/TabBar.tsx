import React, { PureComponent, cloneElement } from 'react';
import classnames from 'classnames';
import type { BaseTabBarProps } from './PropsType';
import type TabBarItem from './TabBarItem';

export interface TabBarProps extends BaseTabBarProps {
  prefixCls?: string;
  className?: string;
}

class TabBar extends PureComponent<TabBarProps, any> {
  static Item: typeof TabBarItem;

  static defaultProps: TabBarProps = {
    prefixCls: 'za-tab-bar',
    visible: true,
  };

  onChildChange = (value: string | number) => {
    const { onChange } = this.props;
    if (typeof onChange === 'function') {
      onChange(value);
    }
  };

  getSelected = (index: number, itemKey: string | number) => {
    const { activeKey, defaultActiveKey } = this.props;
    if (!activeKey) {
      if (!defaultActiveKey && index === 0) {
        return true;
      }
      return defaultActiveKey === itemKey;
    }
    return activeKey === itemKey;
  };

  render() {
    const { visible, prefixCls, className, children, style } = this.props;
    const cls = classnames(prefixCls, className, {
      [`${prefixCls}--hidden`]: !visible,
    });
    const items = React.Children.map(children, (element, index) => {
      if (!React.isValidElement(element)) return null;
      const itemKey = element.props.itemKey || index;
      return cloneElement(element, {
        key: index,
        disabled: element.props.disabled,
        onChange: () => this.onChildChange(itemKey),
        badge: element.props.badge,
        title: element.props.title,
        icon: element.props.icon,
        itemKey,
        style: element.props.style,
        selected: this.getSelected(index, itemKey),
      });
    });
    return (
      <div className={cls} style={style}>
        {items}
      </div>
    );
  }
}

export default TabBar;
