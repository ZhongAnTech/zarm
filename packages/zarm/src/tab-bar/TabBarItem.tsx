import React from 'react';
import classnames from 'classnames';
import Badge from '../badge';
import type { BaseTabBarItemProps } from './interface';
import { ConfigContext } from '../n-config-provider';

export interface TabBarItemProps
  extends BaseTabBarItemProps,
    Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange' | 'title'> {}

const TabBarItem = React.forwardRef<unknown, TabBarItemProps>((props, ref) => {
  const tabBaItemrRef = (ref as any) || React.createRef<HTMLDivElement>();

  const { prefixCls: globalPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-tab-bar`;

  const {
    title,
    icon,
    badge,
    selected,
    activeIcon = icon,
    onChange,
    itemKey,
    ...restProps
  } = props;

  const cls = classnames(`${prefixCls}__item`, {
    [`${prefixCls}--active`]: selected,
  });

  const contentRender = (
    <>
      <div className={`${prefixCls}__icon`}>{selected ? activeIcon : icon}</div>
      <div className={`${prefixCls}__title`}>{title}</div>
    </>
  );

  return (
    <div className={cls} ref={tabBaItemrRef} onClick={onChange} {...restProps}>
      {badge ? <Badge {...badge}>{contentRender}</Badge> : contentRender}
    </div>
  );
});

TabBarItem.displayName = 'TabBarItem';

export default TabBarItem;
