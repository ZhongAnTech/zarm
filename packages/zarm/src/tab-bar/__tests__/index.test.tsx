import React from 'react';
import { render, mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TabBar from '../index';
import TabBarItem from '../TabBarItem';
import Badge from '../../badge';

const activeIcon = (
  <div
    style={{
      width: '24px',
      height: '24px',
      background:
        'url(//cdn-health.zhongan.com/zarm/home-active.svg) top left / 24px 24px no-repeat',
    }}
  />
);

const icon = (
  <div
    style={{
      width: '24px',
      height: '24px',
      background: 'url(//cdn-health.zhongan.com/zarm/home.svg) top left / 24px 24px no-repeat',
    }}
  />
);

describe('TabBarItem', () => {
  it('should render with a badge', () => {
    const wrapper = shallow(<TabBarItem itemKey="1" badge={{ theme: 'primary' }} />);
    expect(wrapper.prop('className')).toBe('za-tab-bar__item');
    expect(wrapper.prop('style')).toBeUndefined();
    expect(wrapper.find(Badge)).toBeTruthy();
    expect(wrapper.find(Badge).prop('theme')).toEqual('primary');
  });

  it('should render with content', () => {
    const wrapper = shallow(
      <TabBarItem itemKey="1" title={<h1>test</h1>} icon={<span>test icon</span>} />,
    );
    expect(wrapper.find('.za-tab-bar__icon')).toBeTruthy();
    expect(wrapper.find('.za-tab-bar__icon').children()).toHaveLength(1);
    expect(wrapper.find('.za-tab-bar__icon').children().at(0).text()).toBe('test icon');
    expect(wrapper.find('.za-tab-bar__icon').children().at(0).type()).toBe('span');
    expect(wrapper.find('.za-tab-bar__title')).toBeTruthy();
    expect(wrapper.find('.za-tab-bar__title').contains(<h1>test</h1>)).toBeTruthy();
  });

  it('should handle change event', () => {
    const mOnChange = jest.fn();
    const wrapper = shallow(<TabBarItem itemKey={1} onChange={mOnChange} />);
    wrapper.simulate('click');
    expect(mOnChange).toBeCalledWith(1);
  });

  it('should render with icon if it is selected', () => {
    const wrapper = shallow(
      <TabBarItem itemKey="1" selected title={<h1>test</h1>} icon={<span>test icon</span>} />,
    );
    expect(wrapper.prop('className')).toEqual('za-tab-bar__item za-tab-bar--active');
    expect(wrapper.find('.za-tab-bar__icon').contains(<span>test icon</span>)).toBeTruthy();
  });

  it('should render with active icon if it is selected', () => {
    const wrapper = shallow(
      <TabBarItem itemKey="1" selected title={<h1>test</h1>} icon={<span>active icon</span>} />,
    );
    expect(wrapper.find('.za-tab-bar__icon').contains(<span>active icon</span>)).toBeTruthy();
  });

  it('should render inline style', () => {
    const wrapper = shallow(<TabBarItem itemKey="1" style={{ color: 'red', display: 'flex' }} />);
    expect(wrapper.prop('style')).toHaveProperty('color', 'red');
    expect(wrapper.prop('style')).toHaveProperty('display', 'flex');
  });
});

describe('TabBar', () => {
  describe('snapshot', () => {
    it('should render correctly', () => {
      const wrapper = render(
        <TabBar>
          <TabBar.Item itemKey="home" title="主页" icon={icon} activeIcon={activeIcon} />
        </TabBar>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('with defaultActiveKey', () => {
      const wrapper = render(
        <TabBar defaultActiveKey="home">
          <TabBar.Item itemKey="home" title="主页" icon={icon} activeIcon={activeIcon} />
        </TabBar>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('with activeKey', () => {
      const wrapper = render(
        <TabBar activeKey="home">
          <TabBar.Item itemKey="home" title="主页" icon={icon} activeIcon={activeIcon} />
        </TabBar>,
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  it('should handle change event', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <TabBar defaultActiveKey="home" onChange={onChange}>
        <TabBar.Item itemKey="home" title="主页" icon={icon} activeIcon={activeIcon} />
      </TabBar>,
    );

    wrapper.find('.za-tab-bar__item').first().simulate('click');
    expect(onChange).toBeCalledWith('home');
  });

  it('should select first tab bar item if activeKey and defaultActivceKey are not existed', () => {
    const wrapper = shallow(
      <TabBar visible={false} className="test">
        <TabBar.Item itemKey="home" title="主页" />
        <TabBar.Item itemKey="about us" title="关于我们" />
      </TabBar>,
    );
    expect(wrapper.childAt(0).prop('selected')).toBeTruthy();
    expect(wrapper.childAt(1).prop('selected')).toBeFalsy();
  });

  it('should render icon if selected is false(defaultActivceKey is not equal with itemKey)', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <TabBar defaultActiveKey="home" onChange={onChange}>
        <TabBar.Item
          itemKey="badge"
          title="主页"
          icon={icon}
          activeIcon={activeIcon}
          badge={{ shape: 'circle', text: '3' }}
        />
      </TabBar>,
    );
    wrapper.find(TabBar.Item).first().simulate('click');
    expect(onChange).toBeCalledWith('badge');
    expect(wrapper.find('.za-tab-bar__icon').contains(icon)).toBeTruthy();
  });

  it('should selected tab bar item if activeKey is equal with itemKey', () => {
    const wrapper = shallow(
      <TabBar defaultActiveKey="home">
        <TabBar.Item itemKey="home" title="主页" />
      </TabBar>,
    );
    expect(wrapper.find(TabBar.Item).prop('selected')).toBeTruthy();
  });

  it('should use element index as the fallback itemKey', () => {
    const wrapper = shallow(
      <TabBar>
        <TabBar.Item />
        <TabBar.Item />
      </TabBar>,
    );
    expect(wrapper.childAt(0).prop('itemKey')).toBe(0);
    expect(wrapper.childAt(1).prop('itemKey')).toBe(1);
  });

  it('should have hidden class name if visible prop is false', () => {
    const wrapper = shallow(
      <TabBar visible={false} className="test">
        <TabBar.Item itemKey="home" title="主页" />
      </TabBar>,
    );
    expect(wrapper.hasClass('test')).toBeTruthy();
    expect(wrapper.hasClass('za-tab-bar')).toBeTruthy();
    expect(wrapper.hasClass('za-tab-bar--hidden')).toBeTruthy();
  });

  it('should render children with extra props', () => {
    const wrapper = shallow(
      <TabBar visible={false} className="test">
        <TabBar.Item itemKey="home" title="主页" />
        <TabBar.Item itemKey="about us" title="关于我们" />
      </TabBar>,
    );
    expect(wrapper.childAt(0).props()).toEqual(
      expect.objectContaining({
        title: '主页',
        itemKey: 'home',
        onChange: expect.any(Function),
        selected: true,
      }),
    );
    expect(wrapper.childAt(1).props()).toEqual(
      expect.objectContaining({
        title: '关于我们',
        itemKey: 'about us',
        onChange: expect.any(Function),
        selected: false,
      }),
    );
  });
});
