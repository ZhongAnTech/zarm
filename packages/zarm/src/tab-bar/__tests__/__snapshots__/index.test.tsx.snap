// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar snapshot should render correctly 1`] = `
<div
  class="za-tab-bar"
>
  <div
    class="za-tab-bar__item za-tab-bar--active"
  >
    <div
      class="za-tab-bar__icon"
    >
      <div
        style="width:24px;height:24px;background:url(//cdn-health.zhongan.com/zarm/home-active.svg) top left / 24px 24px no-repeat"
      />
    </div>
    <div
      class="za-tab-bar__title"
    >
      主页
    </div>
  </div>
</div>
`;

exports[`TabBar snapshot with activeKey 1`] = `
<div
  class="za-tab-bar"
>
  <div
    class="za-tab-bar__item za-tab-bar--active"
  >
    <div
      class="za-tab-bar__icon"
    >
      <div
        style="width:24px;height:24px;background:url(//cdn-health.zhongan.com/zarm/home-active.svg) top left / 24px 24px no-repeat"
      />
    </div>
    <div
      class="za-tab-bar__title"
    >
      主页
    </div>
  </div>
</div>
`;

exports[`TabBar snapshot with defaultActiveKey 1`] = `
<div
  class="za-tab-bar"
>
  <div
    class="za-tab-bar__item za-tab-bar--active"
  >
    <div
      class="za-tab-bar__icon"
    >
      <div
        style="width:24px;height:24px;background:url(//cdn-health.zhongan.com/zarm/home-active.svg) top left / 24px 24px no-repeat"
      />
    </div>
    <div
      class="za-tab-bar__title"
    >
      主页
    </div>
  </div>
</div>
`;
